### PR TITLE
Bucket & Function & Passport

### PR DESCRIPTION
**Pull request by tthiscafer**

### CODE REVIEW

File: stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html [hunk 1]
```diff
@@ -41,8 +41,8 @@ <h4 class="mat-h4 bucket-title">
       <h4 mat-subheader><mat-icon>short_text</mat-icon> Describe</h4>
       <mat-divider></mat-divider>
       <mat-list-item>
-        <form id="describeForm" #describeForm="ngForm" class="title-description-form">
-          <mat-form-field>
+        <form id="describeForm" #describeForm="ngForm" class="title-description-form padding-24">
+          <mat-form-field class="field-title">
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html]: Perhaps we should use `mat-form-field.title` css selector

File: stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html [hunk 2]
```diff
@@ -64,7 +64,8 @@ <h4 mat-subheader><mat-icon>short_text</mat-icon> Describe</h4>
               {{ tm?.errors?.maxlength?.requiredLength }} character.
             </mat-error>
           </mat-form-field>
-          <mat-form-field>
+
+          <mat-form-field class="field-description">
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html]: Perhaps we should use `mat-form-field.description` css selector

File: stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html [hunk 3]
```diff
@@ -229,12 +231,14 @@ <h4 mat-subheader><mat-icon>text_fields</mat-icon> Properties</h4>
               <mat-icon *ngIf="propertyKv.value.readOnly" matTooltip="This property is read-only."
                 >format_clear</mat-icon
               >
-            </ng-template>
-
+            </div>
             <span
+              class="input-placer-area"
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html]: You can use `.input-placer-input` class as well.

Comment (thesayyn) [reply stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html]: It is added by input placer by default

File: stacks/spica/src/bucket/pages/bucket-index/bucket-index.component.ts [hunk 4]
```diff
@@ -41,3 +140,15 @@ export class BucketIndexComponent implements OnDestroy {
     this.dispose.next();
   }
 }
+function __indexOf(collection, node) {
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/bucket-index/bucket-index.component.ts]: Please move these functions into class as a private method

File: stacks/spica/src/bucket/pages/bucket-index/bucket-index.component.ts [hunk 5]
```diff
@@ -41,3 +140,15 @@ export class BucketIndexComponent implements OnDestroy {
...
+  return Array.prototype.indexOf.call(collection, node);
+}
+
+function __isTouchEvent(event: MouseEvent | TouchEvent): event is TouchEvent {
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/bucket-index/bucket-index.component.ts]: Please move these functions into class as a private method

File: stacks/spica/src/bucket/pages/bucket-index/bucket-index.component.ts [hunk 6]
```diff
@@ -41,3 +140,15 @@ export class BucketIndexComponent implements OnDestroy {
...
+  return event.type.startsWith("touch");
+}
+
+function __isInsideDropListClientRect(dropList: CdkDropList, x: number, y: number) {
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/bucket-index/bucket-index.component.ts]: Please move these functions into class as a private method

File: stacks/spica/src/bucket/pages/index/index.component.html [hunk 7]
```diff
@@ -122,8 +122,9 @@ <h4 class="mat-h4">
       (matSortChange)="sortChange($event)"
     >
       <ng-container matColumnDef="$$spicainternal_select">
-        <mat-header-cell *matHeaderCellDef>
+        <mat-header-cell class="select-header" *matHeaderCellDef>
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/index/index.component.html]: We should use `mat-header-cell.select` css selector instead

File: stacks/spica/src/bucket/pages/index/index.component.html [hunk 8]
```diff
@@ -122,8 +122,9 @@ <h4 class="mat-h4">
...
           <mat-checkbox
+            class="select-header"
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/index/index.component.html]: We should use `mat-checkbox.select` css selector instead

File: stacks/spica/src/bucket/pages/index/index.component.scss [hunk 9]
```diff
@@ -16,6 +16,12 @@
     }
   }
 }
+.select-header {
```
Comment (thesayyn) [inline stacks/spica/src/bucket/pages/index/index.component.scss]: We should use `mat-header-cell.select` css selector instead

File: stacks/spica/src/function/pages/index/index.component.html [hunk 10]
```diff
@@ -32,8 +32,10 @@ <h4 class="mat-h4">
       </ng-container>
 
       <ng-container matColumnDef="actions">
-        <mat-header-cell class="actions-header" *matHeaderCellDef>Actions</mat-header-cell>
-        <mat-cell class="actions-content" *matCellDef="let function">
+        <mat-header-cell class="action-header" class="actions-header" *matHeaderCellDef
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/index/index.component.html]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/function/pages/index/index.component.scss [hunk 11]
```diff
@@ -14,6 +14,9 @@
       }
     }
   }
+  .action-header {
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/index/index.component.scss]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/function/pages/subscription-index/subscription-index.component.html [hunk 12]
```diff
@@ -25,8 +25,8 @@ <h4 class="mat-h4">
       </ng-container>
 
       <ng-container matColumnDef="actions">
-        <mat-header-cell *matHeaderCellDef>Actions</mat-header-cell>
-        <mat-cell *matCellDef="let element">
+        <mat-header-cell class="action-cell" *matHeaderCellDef>Actions</mat-header-cell>
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/subscription-index/subscription-index.component.html]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/function/pages/subscription-index/subscription-index.component.html [hunk 13]
```diff
@@ -25,8 +25,8 @@ <h4 class="mat-h4">
...
+        <mat-cell class="action-cell" *matCellDef="let element">
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/subscription-index/subscription-index.component.html]: We should use `mat-cell.actions` css selector instead

File: stacks/spica/src/function/pages/subscription-index/subscription-index.component.scss [hunk 14]
```diff
@@ -1,4 +1,8 @@
 :host {
+  .action-cell {
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/subscription-index/subscription-index.component.scss]: We should use `mat-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/identity-index/identity-index.component.scss [hunk 15]
```diff
@@ -0,0 +1,7 @@
+:host {
+  .actions-header,
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/identity-index/identity-index.component.scss]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/identity-settings/identity-settings.component.scss [hunk 16]
```diff
@@ -74,19 +75,25 @@
...
-                  justify-content: space-between;
-                  .mat-form-field {
-                    width: 31.33%;
-                  }
+            ::ng-deep {
+              mat-form-field {
+                width: 30%;
+              }
+            }
+            .property-field {
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/identity-settings/identity-settings.component.scss]: We should use `mat-form-field.property` css selector instead

File: stacks/spica/src/passport/pages/policy-index/policy-index.component.html [hunk 17]
```diff
@@ -33,8 +33,8 @@ <h4 class="mat-h4">
       </ng-container>
 
       <ng-container matColumnDef="actions">
-        <mat-header-cell *matHeaderCellDef>Actions</mat-header-cell>
-        <mat-cell *matCellDef="let element">
+        <mat-header-cell class="action-cell" *matHeaderCellDef>Actions</mat-header-cell>
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/policy-index/policy-index.component.html]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/policy-index/policy-index.component.html [hunk 18]
```diff
@@ -33,8 +33,8 @@ <h4 class="mat-h4">
...
+        <mat-cell class="action-cell" *matCellDef="let element">
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/policy-index/policy-index.component.html]: We should use `mat-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/strategies-add/strategies-add.component.html [hunk 19]
```diff
@@ -51,7 +51,7 @@ <h4 class="mat-h4">
       <input matInput placeholder="Title" [(ngModel)]="strategy.title" />
       <mat-hint>Title will be shown in the Login page.</mat-hint>
     </mat-form-field>
-    <mat-form-field>
+    <mat-form-field class="custom-type-field">
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/strategies-add/strategies-add.component.html]: We should use `mat-form-field.custom-type` css selector instead

File: stacks/spica/src/passport/pages/policy-index/policy-index.component.scss [hunk 20]
```diff
@@ -0,0 +1,6 @@
+:host {
+  .action-cell {
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/policy-index/policy-index.component.scss]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/strategies-add/strategies-add.component.scss [hunk 21]
```diff
@@ -9,4 +10,7 @@
   textarea {
     height: 350px;
   }
+  .custom-type-field {
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/strategies-add/strategies-add.component.scss]: We should use `mat-form-field.cusotm-type` css selector instead

File: stacks/spica/src/passport/pages/strategies/strategies.component.html [hunk 22]
```diff
@@ -34,8 +34,8 @@ <h4>
       </ng-container>
 
       <ng-container matColumnDef="actions">
-        <mat-header-cell *matHeaderCellDef>Actions</mat-header-cell>
-        <mat-cell *matCellDef="let element">
+        <mat-header-cell class="action-cell" *matHeaderCellDef>Actions</mat-header-cell>
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/strategies/strategies.component.html]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/strategies/strategies.component.html [hunk 23]
```diff
@@ -34,8 +34,8 @@ <h4>
...
+        <mat-cell class="action-cell" *matCellDef="let element">
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/strategies/strategies.component.html]: We should use `mat-cell.actions` css selector instead

File: stacks/spica/src/passport/pages/strategies/strategies.component.scss [hunk 24]
```diff
@@ -0,0 +1,6 @@
+:host {
+  .action-cell {
```
Comment (thesayyn) [inline stacks/spica/src/passport/pages/strategies/strategies.component.scss]: We should use `mat-header-cell.actions` css selector instead

File: stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html (hunk 3 above)
Comment (tthiscafer) [reply stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html]: doesn't exists  .input-placer-input  class. 
image

Comment (thesayyn) [reply stacks/spica/src/bucket/pages/bucket-add/bucket-add.component.html]: Okay then we can pass this

File: stacks/api/bucket/bucket.controller.ts [hunk 25]
```diff
@@ -59,6 +60,17 @@ export class BucketController {
     return this.bs.replaceOne(bucket).then(() => bucket);
   }
 
+  @Put()
+  @UseGuards(AuthGuard(), ActionGuard("bucket:update"))
+  updateMany(@Body(Schema.validate("http://spica.internal/buckets/schema")) buckets: Bucket[]) {
+    return Promise.all(
+      buckets.map(bucket => {
+        bucket._id = new ObjectId(bucket._id);
+        this.bs.updateOne(bucket);
```
Comment (thesayyn) [inline stacks/api/bucket/bucket.controller.ts]: this line should be `return this.bs.updateOne(bucket);`

File: stacks/spica/src/function/pages/index/index.component.scss [hunk 26]
```diff
@@ -14,6 +14,12 @@
       }
     }
   }
+  mat-header-cell .action {
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/index/index.component.scss]: Use `mat-header-cell.action` instead

File: stacks/spica/src/function/pages/subscription-index/subscription-index.component.scss [hunk 27]
```diff
@@ -1,4 +1,10 @@
 :host {
+  mat-header-cell .action {
```
Comment (thesayyn) [inline stacks/spica/src/function/pages/subscription-index/subscription-index.component.scss]: Use `mat-header-cell.action` instead

**Pull request merged**